### ISSUE TITLE
Additional output, e.g. pyppeteer-chromium?

### ISSUE THREAD

**Issue by bollwyvl** [conda-forge/pyppeteer-feedstock#3]
The download of pyppeteer to a global location seems fragile. Because the compatibility window is rather specific, having two versions installed in different envs could lead to really mysterious problems.

I'm wondering if this used `outputs` to generate:
- the `noarch` package, setting the default `PYPPETEER_HOME` with either a patch or an activate script (ick) to point to...
- a non-`noarch` package, `pyppeteer-with-chromium`, with a hard pin on the `noarch` version, that already contained the downloaded binary, so that it doesn't require the network at run time

If you just install `pyppeteer`, the behavior would download/install it to your env, rather than to `$HOME`, while if you used `pyppeteer-with-chromium`, when it imports, it would already have the binary, and you'd be guaranteed to get something that had been tested.

**Comment by kiwi0fruit** [comment]
What particular problems do you see with global chromium download location? I'm only aware of inability to uninstall properly and hence potential memory issues. So if to fix something it' s to move download location into env itself. And I'm not going to package Chromium or upate it. Ever.

**Comment by bollwyvl** [comment]
Sorry for the delay in responding.

> move download location into env itself

Right. The quick patch would be [this line](https://github.com/pyppeteer/pyppeteer/blob/2943975574ac47d3735bad1a70030683918aede5/pyppeteer/__init__.py#L16-L17), just injecting the $PREFIX (through whatever means). 

>  going to package Chromium

But each release of _this_ package specifies a very specific revision of its chromium, but still leaves every end user downloading the uncharacterized binary blob. Warming up the cache during CI, once per platform per version, and packaging that, would save an end user doing their install,  starting a browser test/task and trigger the x00 mb download. Of course the env var should still be honored, and if someone _did_ specify another revision, then they'd have to wait for that to download, but if using conda for something, I want it to do as much of the heavy lifting as possible.

We've kinda done this on https://github.com/conda-forge/python-chromedriver-binary-feedstock, aside from actually using the upstream published binaries, rather than relying on the package itself to do the downloading, so we at least have some traceability/hashes. But there, the bit of needed information to build the download URL is available inside the version string, which is not the case here.

>  or upate it. Ever.

Adopting this approach would be the same number of clicks to maintain, though it would be doing more builds per PR, which means more opportunities for flake. Or, a second feedstock altogether could be created, as we have done for https://github.com/conda-forge/spacy-model-en_core_web-feedstock, keeping this one simple-as-pie noarch... the downstream would rely on this package's patched location being appropriate. But keeping them together would allow for _testing_ them together.

**Comment by kiwi0fruit** [comment]
But at the same time moving download place into the env would make reinstallation of the env to download chromium again. That is worse than memory issues in my opinion. Putting download dir into root conda dir would be a better option but is there a reliable way to learn root environment dir (and may be write it via a post intall conda script so that pyppeteer python script could be aware of it)?

As about conda packaging chromium - I would only consider this after there would be conda forge package with chromium so that pyppeteer could simply depend on it.

**Comment by kiwi0fruit** [comment]
Another option would be to creare a separate conda package that downloads chromium to itself - without packaging chromium. Presumably downloading to itself would affect all environments with this package because conda links packsges instead of copying files.

**Comment by bollwyvl** [comment]
> conda forge package with chromium

Indeed, there have been [several](https://github.com/conda-forge/staged-recipes/pull/7146) [attempts](https://github.com/conda-forge/staged-recipes/pull/5256) to build chromium from source, but i don't think any free CI can handle that beast. QtWebEngine is probably the closest, and is pretty hard. But for API-level integration, as with all things chrom*, the versions have to match extremely closely these days.

On [firefox](https://github.com/conda-forge/firefox-feedstock) we download the binaries and repack them: sure, they have native deps, but it moves the somewhat closer to the goal.

Another compelling reason to have something "in-package, in-env" is you can use [constructor](https://github.com/conda/constructor) to include a browser in an installer. In fact we do this on [robotlab](https://github.com/robots-from-jupyter/robotlab), and aside from weird CI issues, it delivers people something we _know_ will work to learn web automation, because we can test it.

>  conda links packsges instead of copying files.

Yeah, this is the behavior I'm counting on: once it's in e.g. `$BASE/pkgs/pyppeteer-chromium-0.0.25-py37abasdav_1`, so long as as a user is using that version, on the same physical volume, the nominal cost is almost 0: no download, no unpacking.

**Comment by kiwi0fruit** [comment]
I guess repackaging of already built chromium for conda is also OK. Pandoc for example is simply repackaged. But I won't do the repackaging thing.

As about dedicated conda package with hardcoded chromium version that would download chromium of this particular version into itself in the base env pkgs - that sounds nice.

But I guess there are some contradictions with default behavior of pyppeteer that if specified another version would download that specified version to user data. And downloading if specified to extracted conda package is a bad idea...

**Comment by kiwi0fruit** [comment]
> As about dedicated conda package with hardcoded chromium version that would download chromium of this particular version into itself in the base env pkgs - that sounds nice.

But I'm still of the opinion that

> Putting download dir into root conda dir would be a better option but is there a reliable way to learn root environment dir?

is a better option.

If to create a separate package that might be a small package with chromium versions for 3 OSes. Patched Pyppeteer would depend on this package and read Chromium version there and use it by default. Hence instead of updating pyppeteer package itself this small package can be updated instead.

But still it's not that better than updating Chromium version in the pyppeteer package itself.

So the only interesting thing to implement is to move pyppeteer download dir to conda root. I guess conda is aware of it.

**Comment by bollwyvl** [comment]
It seems conda packages should not know about the existence of any environment except the one they are installed in... hard links should already do best-effort deduplication of `$base/pkgs/pyppeteer-download-0.2.2/chromium.tgz` (or whatever). If you're on different file systems :shrug: Better, of course, is if it's already the _usable_ chromium, as unpacking an 80mb (or whatever) tarball isn't free.
